tests/test_flow/test_reload.py

```diff
@@ -0,0 +1,258 @@
+import pytest
+
+import sys
+from importlib import import_module
+from textwrap import dedent
+
+from bionic.exception import CodeVersioningError, UndefinedEntityError
+
+
+@pytest.fixture
+def tmp_module_path(tmp_path):
+    tmp_dir_name = str(tmp_path)
+    sys.path.insert(0, tmp_dir_name)
+
+    yield tmp_path
+
+    # Ensure that the temporary module is not loaded by a future test.
+    sys.path.remove(tmp_dir_name)
+
+
+@pytest.fixture
+def write_module(tmp_module_path):
+    # Pytest rewrites modules and caches the compiled pyc file. The cache can
+    # interfere with the tests below. Hence, we disable the cache here.
+    dont_write_bytecode = sys.dont_write_bytecode
+    sys.dont_write_bytecode = True
+
+    modules = set()
+
+    def _write_module(name, contents):
+        (tmp_module_path / f"{name}.py").write_text(dedent(contents))
+        modules.add(name)
+
+    yield _write_module
+
+    # Invalidate caches
+    for m in modules:
+        sys.modules.pop(m, None)
+
+    sys.dont_write_bytecode = dont_write_bytecode
+
+
+@pytest.fixture
+def write_flow_module(write_module):
+    def _write_flow_module(x):
+        write_module(
+            "flow_module",
+            f"""
+                import bionic as bn
+
+                builder = bn.FlowBuilder('flow')
+                builder.assign('x', {x})
+
+                @builder
+                def x_doubled(x):
+                    return x*2
+
+                flow = builder.build()
+            """,
+        )
+
+    return _write_flow_module
+
+
+@pytest.fixture(params=["reload", "reloading"])
+def reload_method(request):
+    return request.param
+
+
+# Whenever flow.reloading() is called, we get a new copy of the flow with
+# reloaded modules. We save the flow instance before the method is called,
+# and verify that it gives the same result after.
+@pytest.fixture
+def flow_tester(reload_method):
+    expected_results_by_flow = {}
+
+    def _verify_flow_contents(flow, results_map):
+        assert {k: flow.get(k) for k in results_map.keys()} == results_map
+        assert {k: getattr(flow.get, k)() for k in results_map.keys()} == results_map
+
+    class FlowTester:
```

Comment:
I think it's a little clearer and more idiomatic to put all your state and functions in the class, rather having a mixture of class state (`flow` and `expected_results`) and closure state (`reload_method` and `_verify_flow_contents`).

Reply:
Similarly, it's probably more straightforward to just have an `import_flow` method on the tester object than to require each test to construct this class. (Also, if you have a class called `FlowTester` and a fixture called `flow_tester`, I'd expect the latter to be an instance of the former.)

Reply:
The flow module needs to be written before `import flow_module` works. Hence the `FlowTester` class is constructed inside of the test.

Reply:
What I mean is, it might be clearer to leave the flow un-initialized and then initialize it like this:

```
def test_change_entity_value(write_flow_module, flow_tester):
    write_flow_module(1)
    tester.import_flow()
    tester.expect_values(x=1, x_doubled=2)
```

Normally I would agree that it's not great to require this sort of extra initialization, but in this case I think it makes the test both clearer and shorter. Up to you, though!

Reply:
I tried having `FlowTester` initialized with a pytest fixture. The `test_assist_versioning_mode` test fails in pycharm, but passes in pytest. This is very strange. I am wondering if it is related to howhow modules are loaded / cached differently in pycharm compared to pytest. One effect of having `FlowTester` be part of a pytest fixture is that it changes the scope of the `flow_module` module.